prow/pjutil/abort.go

```diff
@@ -0,0 +1,75 @@
+package pjutil
+
+import (
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+	prowapi "k8s.io/test-infra/prow/apis/prowjobs/v1"
+)
+
+// prowClient a minimalistic prow client required by the aborter
+type prowClient interface {
+	//ReplaceProwJob updates the prow job
+	ReplaceProwJob(string, prowapi.ProwJob) (prowapi.ProwJob, error)
+}
+
+// ProwJobResourcesCleanupFn type for a callback function which it is expected to clean up
+// all k8s resources associated with the given prow job
+type ProwJobResourcesCleanupFn func(pj prowapi.ProwJob) error
+
+// ProwJobAborter provides functionality to abort prow jobs
+type ProwJobAborter struct {
+	pjc prowClient
+	log *logrus.Entry
+}
+
+//NewProwJobAborter creates a new ProwJobAborter
+func NewProwJobAborter(pjc prowClient, log *logrus.Entry) *ProwJobAborter {
+	return &ProwJobAborter{
+		log: log,
+		pjc: pjc,
+	}
+}
+
+// TerminateOlderPresubmitJobs aborts all presubmit jobs from the given list that have a newer version. It calls
+// the cleanup callback for each job before updating its status as aborted.
+func (a *ProwJobAborter) TerminateOlderPresubmitJobs(pjs []prowapi.ProwJob, cleanup ProwJobResourcesCleanupFn) error {
+	dupes := make(map[string]int)
+	for i, pj := range pjs {
+		if pj.Complete() || pj.Spec.Type != prowapi.PresubmitJob {
+			continue
+		}
+		n := fmt.Sprintf("%s %s/%s#%d", pj.Spec.Job, pj.Spec.Refs.Org, pj.Spec.Refs.Repo, pj.Spec.Refs.Pulls[0].Number)
```

Comment:
Can we use a type for this and map on it?

```go
type jobIdentifier struct {
    org, repo, job string
    pull int
}

dupes := map[jobIdentifier]int{}
```

Reply:
Defined a dedicated type for job identifier.